lib/bundle/installer.rb

```diff
@@ -5,47 +5,121 @@ module Installer
     module_function
 
     def install(entries)
-      success = 0
-      failure = 0
-
-      entries.each do |entry|
-        arg = [entry.name]
-        verb = "Installing"
-        cls = case entry.type
-        when :brew
-          arg << entry.options
-          Bundle::BrewInstaller
-        when :cask
-          arg << entry.options
-          Bundle::CaskInstaller
-        when :mac_app_store
-          arg << entry.options[:id]
-          Bundle::MacAppStoreInstaller
-        when :tap
-          verb = "Tapping"
-          arg << entry.options
-          Bundle::TapInstaller
-        end
-        case cls.install(*arg)
-        when :success
-          puts Formatter.success("#{verb} #{entry.name}")
-          success += 1
-        when :skipped
-          puts "Using #{entry.name}"
-          success += 1
+      end_state = install_entries(entries)
+
+      failures = end_state[:failed]
+      report_success_and_failure(end_state)
+
+      errored_entries = end_state[:errored]
+      report_errors(errored_entries) if errored_entries.any?
+
+      failures.zero? && errored_entries.empty?
+    end
+
+    def task_for_entry(entry)
+      # TODO: move verb into the Installer classes
+      verb = 'Installing'
+      arg = [entry.name]
+      cls = case entry.type
+            when :brew
+              arg << entry.options
+              Bundle::BrewInstaller
+            when :cask
+              arg << entry.options
+              Bundle::CaskInstaller
+            when :mac_app_store
+              arg << entry.options[:id]
+              Bundle::MacAppStoreInstaller
+            when :tap
+              verb = 'Tapping'
+              arg << entry.options
+              Bundle::TapInstaller
+            end
+      InstallerTask.new cls, arg, verb, entry
+    end
+
+    def report_result(task, result)
+      case result
+      when :success
+        puts Formatter.success(task.header)
+      when :skipped
+        puts "Using #{task.entry.name}"
+      when :failed
+        puts Formatter.error("#{task.header} has failed!")
+      else
+        puts Formatter.error("#{task.header} in unknown state: #{result}")
+      end
+    end
+
+    def report_error(task, error)
+      puts Formatter.error("#{task.header} raised an exception: #{error}")
+    end
+
+    def execute_task(task)
+      begin
+        result = task.cls.install(*task.arg)
+        report_result(task, result)
+        { result => 1 }
```

Comment:
It's unclear what `1` means in this context.

Reply:
The return of this is getting merged into the status hash. It's an unclear vestige that will go away if I move away from the status Int|Hash hash as it currently exists on lines 70-75.